velox/functions/prestosql/DateTimeFunctions.h

```diff
@@ -83,6 +83,41 @@ struct TimestampWithTimezoneSupport {
 
 } // namespace
 
+template <typename T>
+struct DateFunction : public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE int64_t getDays(const Timestamp timestamp) {
+    auto seconds = timestamp.getSeconds();
+    if (seconds >= 0 || seconds % kSecondsInDay == 0) {
+      return seconds / kSecondsInDay;
+    }
+    // For division with negatives, minus 1 to compensate the discarded
+    // fractional part. e.g. -1/86'400 yields 0, yet it should be considered as
+    // -1 day.
+    return seconds / kSecondsInDay - 1;
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Varchar>& date) {
+    bool nullOutput;
+    result = util::Converter<TypeKind::DATE>::cast(date, nullOutput);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Timestamp>& timestamp) {
+    result = Date(getDays(timestamp));
```

Comment:
I believe it's a bug.
This formula works fine for non-negative numbers:
| seconds | seconds / 86'400 | day |
|---------|------------------|-----|
| 0       | 0                | 0   |
| 1       | 0                | 0   |
| 86'399  | 0                | 0   |
| 86'400  | 1                | 1   |
| 86'401  | 1                | 1   |

However, for the negatives it may convert the second to the wrong day:
| seconds  | seconds / 86'400 | day    |
|----------|------------------|--------|
| -1       | **0**            | **-1** |
| -86'399  | **0**            | **-1** |
| -86'400  | -1               | -1     |
| -86'401  | **-1**           | **-2** |
| -172'802 | -2               | -2     |

I have updated the Cast implementation accordingly.